container/tabs.go

```diff
@@ -0,0 +1,694 @@
+}
+
+func (r *tabButtonRenderer) Destroy() {
+}
+
+func (r *tabButtonRenderer) Layout(size fyne.Size) {
+	r.background.Resize(size)
+	padding := r.padding()
+	innerSize := size.Subtract(padding)
+	innerOffset := fyne.NewPos(padding.Width/2, padding.Height/2)
+	labelShift := float32(0)
+	if r.icon != nil {
+		var iconOffset fyne.Position
+		if r.button.iconPosition == buttonIconTop {
+			iconOffset = fyne.NewPos((innerSize.Width-r.iconSize())/2, 0)
+		} else {
+			iconOffset = fyne.NewPos(0, (innerSize.Height-r.iconSize())/2)
+		}
+		r.icon.Resize(fyne.NewSize(r.iconSize(), r.iconSize()))
+		r.icon.Move(innerOffset.Add(iconOffset))
+		labelShift = r.iconSize() + theme.Padding()
+	}
+	if r.label.Text != "" {
+		var labelOffset fyne.Position
+		var labelSize fyne.Size
+		if r.button.iconPosition == buttonIconTop {
+			labelOffset = fyne.NewPos(0, labelShift)
+			labelSize = fyne.NewSize(innerSize.Width, r.label.MinSize().Height)
+		} else {
+			labelOffset = fyne.NewPos(labelShift, 0)
+			labelSize = fyne.NewSize(innerSize.Width-labelShift, innerSize.Height)
+		}
+		r.label.Resize(labelSize)
+		r.label.Move(innerOffset.Add(labelOffset))
+	}
+	r.close.Move(fyne.NewPos(size.Width-theme.IconInlineSize()-theme.Padding(), (size.Height-theme.IconInlineSize())/2))
+	r.close.Resize(fyne.NewSize(theme.IconInlineSize(), theme.IconInlineSize()))
+}
+
+func (r *tabButtonRenderer) MinSize() fyne.Size {
+	var contentWidth, contentHeight float32
+	textSize := r.label.MinSize()
+	if r.button.iconPosition == buttonIconTop {
+		contentWidth = fyne.Max(textSize.Width, r.iconSize())
+		if r.icon != nil {
+			contentHeight += r.iconSize()
+		}
+		if r.label.Text != "" {
+			if r.icon != nil {
+				contentHeight += theme.Padding()
+			}
+			contentHeight += textSize.Height
+		}
+	} else {
+		contentHeight = fyne.Max(textSize.Height, r.iconSize())
+		if r.icon != nil {
+			contentWidth += r.iconSize()
+		}
+		if r.label.Text != "" {
+			if r.icon != nil {
+				contentWidth += theme.Padding()
+			}
+			contentWidth += textSize.Width
+		}
+	}
+	if r.button.onClosed != nil {
+		contentWidth += theme.IconInlineSize() + theme.Padding()
+		contentHeight = fyne.Max(contentHeight, theme.IconInlineSize())
+	}
+	return fyne.NewSize(contentWidth, contentHeight).Add(r.padding())
+}
+
+func (r *tabButtonRenderer) Objects() []fyne.CanvasObject {
+	return r.objects
+}
+
+func (r *tabButtonRenderer) Refresh() {
+	if r.button.hovered {
+		r.background.FillColor = theme.HoverColor()
+		r.background.Show()
+	} else {
+		r.background.Hide()
+	}
+	r.background.Refresh()
+
+	r.label.Text = r.button.text
+	r.label.Alignment = r.button.textAlignment
+	if r.button.importance == widget.HighImportance {
+		r.label.Color = theme.PrimaryColor()
+	} else {
+		r.label.Color = theme.TextColor()
+	}
+	r.label.TextSize = theme.TextSize()
+	if r.button.text == "" {
+		r.label.Hide()
+	} else {
+		r.label.Show()
+	}
+
+	if r.icon != nil && r.icon.Resource != nil {
+		switch res := r.icon.Resource.(type) {
```

Comment:
This code seems unnecessary as the icons are always shown with the foreground colour...

Reply:
Though now I type it maybe the issue that the button icons are not being changed to primary when they should?

Reply:
Yup, I didn't change this code, it is copied from https://github.com/fyne-io/fyne/blob/develop/container/apptabs.go#L648

Reply:
Well, it's not being called now as far as I can see - the icons are the same colour on selected and not-selected tabs

Reply:
Where are you looking? The demo doesn't use icons, but when I add one it works as expected

https://user-images.githubusercontent.com/1117080/106517086-dd8aeb00-648c-11eb-9e73-ea8b6b196e86.mov

Reply:
Apologies, I forgot there were two - I meant the close icon

Reply:
You want the close icon to be coloured differently if the tab is selected? What colour should it be?

Reply:
What I think I was expecting was the close icon to be the same colour as the tab icon.
Don't you think so?

Reply:
When selected the tab's text and icon become the primary colour, which I feel has an important conveyance about what the main action a user is expected to do. Eg button uses primary colour to convey that it has high importance. Maybe tab item should use selected colour when that is introduced, but that's a separate discussion.

Close is a separate action, and a destructive one at that, so my first thought was that it should be a warning colour like red, that would be used all the time and not depending on whether the tab item is selected, but @fpabl0 didn't feel it was necessary. My main worry is that in the icon-only use case, two icons next to each other with the same colour where one changes the selected tab and the other removes tabs, it would be easy for the user to accidentally close tho I tried to avoid this somewhat by making the close button smaller and inset.

Reply:
> Close is a separate action, and a destructive one at that, so my first thought was that it should be a warning colour like red, that would be used all the time and not depending on whether the tab item is selected

IMHO, it should not be colored, as the intention of the tabs is to be selected not be closed. Usually in ui design, as far as I know, you want to contrast things that user usually want to use. Closing a tab is not the main purpose so it should not have a contrast color in my opinion.

Reply:
For example, online stores have a button to delete an item from the cart, but although it is a destructive action is not red, because the main purpose of the cart is not delete items, but buy them. From amazon: (Delete is "Eliminar" (spanish)):

<img width="909" alt="Captura de Pantalla 2021-02-07 a la(s) 13 50 47" src="https://user-images.githubusercontent.com/12239342/107156339-c3994e80-694b-11eb-9221-be4f66ff2646.png">

Reply:
> so my first thought was that it should be a warning colour like red

This sounds like it could be helpful - we have `theme.ErrorColor` and appropriate resource helper.

> My main worry is that in the icon-only use case, two icons next to each other with the same colour where one changes the selected tab and the other removes tabs, it would be easy for the user to accidentally close

Not sure I fully follow this - but if the close button is white on a tab that has a blue icon (because it is selected) and the icon next to it is white (because it is not selected) then is it not easier to mistake the close icon for belonging to the white icon to the right?

Reply:
(not the best example, but this is VS Code)
<img width="437" alt="Screenshot 2021-02-08 at 20 50 03" src="https://user-images.githubusercontent.com/294436/107279677-40f9b780-6a4f-11eb-8f8c-3d1013239a3c.png">

Reply:
> When I look at the tabs on my computer where the colour of the text changes for active/inactive the majority of the close icons match the colour of the text.

Yes, you are right. But they don't have the primary color on it. The primary color is the background not the text color.

Reply:
> (not the best example, but this is VS Code)

Indeed, it is a good example. That looks good, because the text color is white and not red, it would look weird if it uses the primary color.

Reply:
I thought about using `theme.ErrorColor` but the semantics of using error to convey danger seemed off.

I agree with Pablo that the way the selection is shown is different. Where in our tabbers change the foreground colour but that VS code example change the background colour. Maybe it would be better if tabbers followed the new collections which use background to indicate selection?

Reply:
>  Maybe it would be better if tabbers followed the new collections which use background to indicate selection?

I think it would look better with the background to indicate selection :)

Reply:
It would get rid of the current underline indicator and associated animation

Reply:
Yes, another way could be this:

<img width="641" alt="Captura de Pantalla 2021-02-08 a la(s) 16 52 13" src="https://user-images.githubusercontent.com/12239342/107286378-822d9f80-6a2e-11eb-930d-a14f137c409d.png">

<img width="541" alt="Captura de Pantalla 2021-02-08 a la(s) 16 56 31" src="https://user-images.githubusercontent.com/12239342/107286469-a1c4c800-6a2e-11eb-9e74-fc7757f619bb.png">

However, a new theme color would be needed, I think.

Reply:
> that VS code example change the background colour.

If you look carefully it is changing both

> Maybe it would be better if tabbers followed the new collections which use background to indicate selection?

We used to do that - people did not like it and it is not what material uses for tabs

What I don't follow is why any of these thoughts provide a compelling reason to break from the MD looks.

Reply:
Fair enough, the Tabs spec (https://material.io/components/tabs) doesn't mention anything about closing tabs so I'm inclined to leave this design as is until we know better and can iterate in future.

Reply:
> We used to do that - people did not like it and it is not what material uses for tabs

I see, that's fair. As @stuartmscott said, Material does not mention anything about Doc Tabs (with close button), because it does not exist there, as far as I know. However, after looking the Google Cloud Shell terminal, @andydotxyz suggestion maybe its the better choice?

<img width="435" alt="Captura de Pantalla 2021-02-09 a la(s) 11 29 34" src="https://user-images.githubusercontent.com/12239342/107394668-1ef14b00-6aca-11eb-85ee-2f7f5761a695.png">

Looking at how this is implemented in other apps based on Material Design Specs, I found this:

![image](https://user-images.githubusercontent.com/12239342/107395007-71cb0280-6aca-11eb-99df-6839cb37e943.png)

Reply:
Close buttons now become primary coloured if tab is selected;

https://user-images.githubusercontent.com/1117080/107537202-78638380-6b77-11eb-84f1-e5e7a9b12c4e.mov